src/EventStore.Core.XUnit.Tests/Services/Storage/InMemoryStreamReaderTests.cs

```diff
@@ -0,0 +1,149 @@
+			[SystemStreams.NodeStateStream] = _listener,
+		});
+	}
+
+	public class ReadForwardEmptyTests : InMemoryStreamReaderTests {
+		[Fact]
+		public void read_forwards_empty() {
+			var correlation = Guid.NewGuid();
+			var result = _sut.ReadForwards(
+				new ClientMessage.ReadStreamEventsForward(
+					correlation,
+					correlation,
+					new NoopEnvelope(),
+					SystemStreams.NodeStateStream,
+					0,
+					10,
+					false,
+					false,
+					null,
+					ClaimsPrincipal.Current,
+					true));
+
+			Assert.Equal(correlation, result.CorrelationId);
+			Assert.Equal(ReadStreamResult.NoStream, result.Result);
+			Assert.Equal(SystemStreams.NodeStateStream, result.EventStreamId);
+			Assert.Equal(0, result.FromEventNumber);
+			Assert.Equal(10, result.MaxCount);
+			Assert.Equal(-1, result.NextEventNumber);
+			Assert.Equal(-1, result.LastEventNumber);
+			Assert.Empty(result.Events);
+		}
+	}
+
+	public class ReadForwardTests : InMemoryStreamReaderTests {
+		[Fact]
+		public void read_forwards() {
+			_listener.Handle(new SystemMessage.BecomeLeader(Guid.NewGuid()));
+			var correlation = Guid.NewGuid();
+			var result = _sut.ReadForwards(
+				new ClientMessage.ReadStreamEventsForward(
+					correlation,
+					correlation,
+					new NoopEnvelope(),
+					SystemStreams.NodeStateStream,
+					0,
+					10,
+					false,
+					false,
+					null,
+					ClaimsPrincipal.Current,
+					true));
+
+			Assert.Equal(correlation, result.CorrelationId);
+			Assert.Equal(ReadStreamResult.Success, result.Result);
+			Assert.Equal(SystemStreams.NodeStateStream, result.EventStreamId);
+			Assert.Equal(0, result.FromEventNumber);
+			Assert.Equal(10, result.MaxCount);
+			Assert.Equal(1, result.NextEventNumber);
+			Assert.Equal(0, result.LastEventNumber);
+			Assert.Single(result.Events);
+
+			var @event = result.Events[0];
+			Assert.Equal(SystemStreams.NodeStateStream, @event.Event.EventStreamId);
+			Assert.Equal(NodeStateListenerService.EventType, @event.Event.EventType);
+		}
+	}
+
+	public class ReadBackwardsEmptyTests : InMemoryStreamReaderTests {
+		[Fact]
+		public void read_backwards_empty() {
+			var correlation = Guid.NewGuid();
+			var result = _sut.ReadBackwards(
+				new ClientMessage.ReadStreamEventsBackward(
+					correlation,
+					correlation,
+					new NoopEnvelope(),
+					SystemStreams.NodeStateStream,
+					0,
+					10,
+					false,
+					false,
+					null,
+					ClaimsPrincipal.Current));
+
+			Assert.Equal(correlation, result.CorrelationId);
+			Assert.Equal(ReadStreamResult.NoStream, result.Result);
+			Assert.Equal(SystemStreams.NodeStateStream, result.EventStreamId);
+			Assert.Equal(0, result.FromEventNumber);
+			Assert.Equal(10, result.MaxCount);
+			Assert.Equal(-1, result.NextEventNumber);
+			Assert.Equal(-1, result.LastEventNumber);
+			Assert.Empty(result.Events);
+		}
+	}
+
+	public class ReadBackwardsTests : InMemoryStreamReaderTests {
+		[Fact]
+		public void read_backwards() {
```

Comment:
Would it be worth adding tests for reading forward/backward from a position greater/less than the last node status event position?

Reply:
Done